initial implementation of search in ui_next

jlmitch5 opened the pull request:
This is the implementation of search in the new patternfly/react-based UI of awx.

All information and documentation about the search feature can be found at https://github.com/jlmitch5/awx/blob/uiNextSearch/awx/ui_next/SEARCH.md

In this pr:
- DONE update handleSearch to follow handleSort param
- DONE update qsConfig columns to utilize isSearchable bool (just like isSortable bool)
- DONE enter keydown in text search bar to search
- DONE get decoded params and write test
- DONE make list header component
- DONE make filter component
- DONE make filters show up for empty list
- DONE make clear all button
- DONE styling of FilterTags component
- DONE clear out text input after tag has been made
- DONE deal with duplicate key tags being added/removed in qs util file
- DONE deal with widgetry changing between one dropdown option to the left of search and many
- DONE bug: figure out why ?name=org returning just org not “org 2”
- DONE update contrib file to have the first section with updated text as is in this pr description.
- DONE rebase with latest awx-pf changes
- DONE styling of search bar
- DONE make filter and list header tests
- DONE change api paramsSerializer to handle duplicate key stuff
- DONE update qs update function to be smaller, simple param functions, as opposed to one big one with a lot of params
- DONE add search filter removal test for qs.
- DONE bug: remove button for search tags of duplicate keys are broken, fix that

Later on in 3.6, the following should be finished for search iteration 1 (I'll card up an issue to tackle this)
- currently handleSearch in Search.jsx always appends the __icontains post-fix to make the filtering ux expected work right. Once we start adding number-based params we will won't to change this behavior.
- utilize new defaultSearchKey prop instead of relying on sort key
- make access have username as the default key?
- make default params only accept page, page_size and order_by
- support custom order_by being typed in the url bar
- fix up which keys are displayed in the various lists (note this will also require non-string widgetry to the right of the search key dropdown, for integers, dates, etc.)
- fix any spacing issues like collision with action buttons and overall width of the search bar

Commit:
initial implementation of search in UI_next

jakemcdermott commented on awx/ui_next/src/components/FilterTags/FilterTags.jsx:
```diff
@@ -0,0 +1,80 @@
+`;
+
+// remove non-default query params so they don't show up as filter tags
+const filterOutDefaultParams = (paramsArr, config) => {
```
How do you feel about naming this`filterDefaultParams` instead?

jakemcdermott commented on awx/ui_next/src/components/FilterTags/FilterTags.jsx:
```diff
@@ -0,0 +1,80 @@
+        {`${itemCount} results`}
+      </ResultCount>
+      <VerticalSeparator />
+      <FilterLabel>Active Filters:</FilterLabel>
```
need some i18n here

jakemcdermott commented on awx/ui_next/src/components/ListHeader/ListHeader.jsx:
```diff
@@ -0,0 +1,140 @@
+    return [queryParams.order_by, 'ascending'];
+  }
+
+  handleSearch (key, value) {
```
It would be nice if `handleSearch` and `handleRemove` had the same call signature - they both accept the same information (key, value) but one expects an object while the other expects an argument list.

jakemcdermott commented on awx/ui_next/src/components/PaginatedDataList/PaginatedDataList.jsx:
```diff
@@ -4,99 +4,54 @@ import { DataList } from '@patternfly/react-core';
+import DataListToolbar from '@components/DataListToolbar';
+
 import {
-  parseNamespacedQueryString,
```
:+1:  I like the changes here because they:
   - maintain the existing external contract between paginated data list and components that use it
   - simplify the _implementation_ of paginated data list by refactoring it into smaller, more testable components.

keithjgrant replied:
Agreed. This is a nice improvement!

jakemcdermott commented on awx/ui_next/src/components/PaginatedDataList/PaginatedDataList.jsx:
```diff
@@ -111,76 +66,61 @@ class PaginatedDataList extends React.Component {
     if (hasContentLoading && items.length <= 0) {
-      Content = <ContentLoading />;
-    } else if (contentError) {
-      Content = <ContentError error={contentError} />;
```
We want to keep passing the whole `contentError` object to the `ContentError` components here - this looks like it might be an unintentional change from resolving this PR with recent changes to error handling that occurred upstream while you were working on this. The content error components now show expandable error details, so they expect the whole error object now.

jakemcdermott commented on awx/ui_next/src/components/PaginatedDataList/PaginatedDataList.test.jsx:
```diff
@@ -88,16 +51,16 @@ describe('<PaginatedDataList />', () => {
     wrapper.update();
     pagination.prop('onSetPage')(null, 1);
-    expect(history.location.search).toEqual('?item.page=1');
+    // since page = 1 is the default, that should be strip out of the search
```
I think we should leave the defaults in the urls for two reasons:

1. Users save and bookmark filtered search results so they can come back to them. By eliding some params from the url, we potentially change bookmarked search results if we ever decide to change any of these default params. I've had this happen several times after upgrades, etc.

2. Prettifying the url, omitting defaults, etc. is a nice to have but it's also a significant contributor of complexity in our current implementation because it impedes our ability to rely on the URL as a source of truth for filtering. I'm not necessarily against doing it here but in general we should proceed with caution in terms of how much of this sort of thing we commit to supporting.

jlmitch5 replied:
We discussed this offline.  Attempted to see if we could use the same encode qs function everywhere (for api payload and for ui urls), but due to the fact we need to namespace the params to put them url (and not for api), we need to have a qsConfig-aware function.  As such it's only a few extra lines needed to filter those defaults out, we've decided to keep the functionality (and those scenarios are covered in the unit tests).

jakemcdermott commented on awx/ui_next/src/components/Search/Search.jsx:
```diff
@@ -65,78 +77,107 @@ class Search extends React.Component {
     const { onSearch } = this.props;
 
-    onSearch(searchValue);
+    // TODO: probably not _always_ add icontains.  I don't think icontains works for numbers.
```
This seems like a good TODO to handle in a follow-up pr. If i'm following the problem correctly, this is where we start to get into how query params might need to vary depending on _which_ field we're adding a filter for.

The `types.js` definitions @keithjgrant has started to put in place may be something we can leverage here. Going forward, our search implementation could use our `types` as a guide for which query operation(s) may be used for a given field.

jakemcdermott commented on awx/ui_next/src/screens/Organization/OrganizationAccess/OrganizationAccess.jsx:
```diff
@@ -32,13 +29,13 @@ class OrganizationAccess extends React.Component {
     super(props);
     this.state = {
       accessRecords: [],
-      contentError: null,
```
We'd want to keep these as `contentError`objects

jakemcdermott reviewed (commented):
awesome :) This is looking like a great start. I left a few comments to get the PR review started. This is a substantial addition of new functionality so I'll revisit this and add more feedback soon.

Also: when you get a moment, please go ahead and run the formatter against this PR and push those changes up, that should resolve the failing zuul test.

jakemcdermott commented on awx/ui_next/src/screens/Organization/OrganizationList/OrganizationList.jsx:
```diff
@@ -23,13 +26,13 @@ const QS_CONFIG = getQSConfig('organization', {
     this.state = {
       hasContentLoading: true,
-      contentError: null,
-      deletionError: null,
```
see https://github.com/ansible/awx/pull/4368/files#r307295389

Commit:
working commit

Commit:
updates based on pr feedback
run prettier
update hasContentError to contentError in all the places
function naming updates

keithjgrant commented on awx/ui_next/src/api/mixins/Notifications.mixin.js:
```diff
@@ -59,8 +60,7 @@ const NotificationsMixin = parent =>
      * @param[notificationType] - the type of notification, options are "success" and "error"
-     * @param[associationState] - Boolean for associating or disassociating,
-     *                            options are true or false
+     * @param[associationState] - Boolean for associating or disassociating, options are true or false
```
Boolean flags in func params can be confusing. It might be worth splitting this into an `associateNotificationTemplate` and `disasociateNotificationTemplate` functions at some point. This would also cut down the number of function arguments.

keithjgrant commented on awx/ui_next/src/api/models/Organizations.js:
```diff
@@ -8,11 +8,11 @@ class Organizations extends InstanceGroupsMixin(NotificationsMixin(Base)) {
 
-  readAccessList(id, params = {}) {
+  readAccessList(id, params) {
     return this.http.get(`${this.baseUrl}${id}/access_list/`, { params });
```
This will create an object like:

```
{
  params: { ... }
}
```

Should this be spread `{ ...params }` instead?

jlmitch5 replied:
this is what seems to work, I think the api or axios expects it on this key.  things on this key run through the paramsSerializer function set up in base.

keithjgrant commented on awx/ui_next/src/components/Chip/ChipGroup.jsx:
```diff
@@ -1,9 +1,15 @@
+  children,
+  className,
+  showOverflowAfter,
+  displayAll,
```
It looks like this will automatically display all chips if `showOverflowAfter` is not given. We may not need the `displayAll` prop here. (I could see an argument that this is confusing though, in which case I would suggest defaulting `showOverFlowAfter` to a reasonable number like 5)

Commit:
update qs addParams/removeParams fns to take param object not string

keithjgrant commented on awx/ui_next/src/util/qs.js:
```diff
@@ -1,54 +1,163 @@
+    denamespaced[key.substr(namespace.length + 1)] = params[key];
+  });
+
+  return denamespaced || {};
```
The `|| {}` seems unnecessary

Commit:
assorted ui_next search phase 1 pr feedback updates
- remove unnecessary displayAll prop from ChipGroup
- update notification api fn to be 2 with no boolean param
- fix params passing to api functions

keithjgrant commented on awx/ui_next/src/util/qs.js:
```diff
@@ -1,54 +1,163 @@
+
+  return isEqual;
+};
+
```
You can probably get rid of `isEqual` var here and just return early in each of the if blocks

jakemcdermott reviewed (approved):
looks good to me, all feedback addressed. That said, this is a significant addition so I think it makes sense to get at least two approvals on this one

Commit:
small updates to qs syntax based on feedback

jlmitch5 commented:
recheck

jlmitch5 commented:
recheck

jakemcdermott commented:
recheck

jakemcdermott commented:
recheck

jlmitch5 committed:
Update SEARCH.md

jlmitch5 committed:
Update SEARCH.md

jakemcdermott commented:
recheck

jakemcdermott commented:
regate
